build.c

```diff
@@ -599,17 +656,30 @@ build(void)
 		}
 		if (numjobs == 0)
 			break;
-		if (poll(fds, jobslen, 5000) < 0)
+		if (poll(fds, 1 + jobslen, 5000) < 0)
 			fatal("poll:");
 		for (i = 0; i < jobslen; ++i) {
-			if (!fds[i].revents || jobwork(&jobs[i]))
+			if (!fds[1+i].revents || jobwork(&jobs[i]))
 				continue;
 			--numjobs;
 			jobs[i].next = next;
-			fds[i].fd = -1;
+			fds[1+i].fd = -1;
 			next = i;
 			if (jobs[i].failed)
 				++numfail;
+			/* return a token */
+			if (isjobserverclient() && gmakelatest > gmaketokens) {
+				if (write(buildopts.gmakepipe[1], --gmakelatest, 1) < 0)
+					warn("write to jobserver:");
+			}
+		}
+		if (fds[0].revents & POLLIN && isjobserverclient() && nstarted < ntotal && !reachedload()) {
+			gmakeread = read(buildopts.gmakepipe[0], gmakelatest, buildopts.maxjobs - numjobs);
```

Comment:
There is a race condition here:

* `poll` returns POLLIN
* another process steals the token that `poll` detected
* `read` blocks
* one or more children of samurai finish
* the whole build cannot reuse the token(s) currently owned by samurai

The solution used by GNU Make is complex and relies on [trapping SIGCHLD and dup-ing the file descriptor](https://make.mad-scientist.net/papers/jobserver-implementation/). It's complicated and it's why I just went for [reading the jobserver file descriptor in a separate thread](https://github.com/michaelforney/samurai/pull/104) instead.

Reply:
I'm not even sure GNU Make's solution is 100% robust without an extra `(void)alarm(1);`, or only starting a new process as an absolute last resort.

I really have to rethink this implementation.